iceoryx2-ffi/cxx/include/iox2/waitset.hpp

```diff
@@ -0,0 +1,222 @@
+
+template <ServiceType S>
+auto operator==(const AttachmentId<S>& lhs, const AttachmentId<S>& rhs) -> bool;
+
+template <ServiceType S>
+auto operator<(const AttachmentId<S>& lhs, const AttachmentId<S>& rhs) -> bool;
+
+/// The [`WaitSet`] implements a reactor pattern and allows to wait on multiple events in one
+/// single call [`WaitSet::run_once()`] until it wakes up or to run repeatedly with
+/// [`WaitSet::run()`] until the a interrupt or termination signal was received or the user
+/// has explicitly requested to stop with [`WaitSet::stop()`].
+///
+/// The [`Listener`] can be attached as well as sockets or anything else that
+/// can be packed into a [`FileDescriptorView`].
+///
+/// Can be created via the [`WaitSetBuilder`].
+template <ServiceType S>
+class WaitSet {
+  public:
+    WaitSet(const WaitSet&) = delete;
+    auto operator=(const WaitSet&) -> WaitSet& = delete;
+    WaitSet(WaitSet&&) noexcept;
+    auto operator=(WaitSet&&) noexcept -> WaitSet&;
+    ~WaitSet();
+
+    /// Can be called from within a callback during [`WaitSet::run()`] to signal the [`WaitSet`]
+    /// to stop running after this iteration.
+    void stop();
+
+    /// Waits in an infinite loop on the [`WaitSet`]. The provided callback is called for every
+    /// attachment that was triggered and the [`AttachmentId`] is provided as an input argument to
+    /// acquire the source.
+    /// If an interrupt- (`SIGINT`) or a termination-signal (`SIGTERM`) was received, it will exit
+    /// the loop and inform the user via [`WaitSetRunResult`].
+    auto run(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<WaitSetRunResult, WaitSetRunError>;
+
+    /// Tries to wait on the [`WaitSet`]. The provided callback is called for every attachment that
+    /// was triggered and the [`AttachmentId`] is provided as an input argument to acquire the
+    /// source.
+    /// If nothing was triggered the [`WaitSet`] returns immediately.
+    auto run_once(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<void, WaitSetRunError>;
+
+    /// Returns the capacity of the [`WaitSet`]
+    auto capacity() const -> uint64_t;
+
+    /// Returns the number of attachments.
+    auto len() const -> uint64_t;
+
+    /// Returns true if the [`WaitSet`] has no attachments, otherwise false.
+    auto is_empty() const -> bool;
+
+    /// Attaches a [`Listener`] as notification to the [`WaitSet`]. Whenever an event is received on the
+    /// object the [`WaitSet`] informs the user in [`WaitSet::run()`] to handle the event.
+    /// The object cannot be attached twice and the
+    /// [`WaitSet::capacity()`] is limited by the underlying implementation.
+    ///
+    /// # Safety
+    ///
+    /// * The [`Listener`] must life at least as long as the returned [`Guard`].
+    /// * The [`Guard`] must life at least as long as the [`WaitsSet`].
+    auto attach_notification(const Listener<S>& listener) -> iox::expected<Guard<S>, WaitSetAttachmentError>;
```

Comment:
Again naming... sorry for being pedantic... But `attach_notification` "Attaches a [`Listener`]" is inconsistent.

The usage seems to be "attach listener" then "get notification".

Reply:
The notification is defined in the rust documentation of the waitset. See:

* **Notification** - An object that emits an event. Whenever the event is detected the
    [`WaitSet`](crate::port::waitset::WaitSet) wakes up and informs the user.
    Typical use case are gateways, which receives and forwards data whenever new data
    is available.
* **Deadline** - Like a *Notification* with the exception that the *Deadline* expects an
    event after a certain predefined timeout. If the event does not arrive before the
    timeout has passed, the [`WaitSet`](crate::port::waitset::WaitSet) wakes up and informs
    the user that th *Deadline* has missed its timeout.
    Whenever a *Deadline* receives an event, the timeout is reset.
    One example is a sensor that shall send an update every 100ms and the applications requires
    the sensor data latest after 120ms. If after 120ms an update
    is not available the application must wake up and take counter measures. If the update
    arrives within the timeout, the timeout is reset back to 120ms.
* **Interval** - An time period after which the [`WaitSet`](crate::port::waitset::WaitSet)
    wakes up and informs the user that the time has passed by.
    This is useful when a [`Publisher`](crate::port::publisher::Publisher) shall send an
    heartbeat every 100ms.

So it shall read that the listener is attached to the waitset as a notification.

Reply:
Is knowing the these details strictly necessary for a user? It seems `Notification` is an internal abstraction for handling the events.

Documentation is good for developers that need to get into the details. My perspective is that if the developer can understand generally what is happening from the API alone, it makes the API easier to use.

My perspective is that the API to describe what the user "wants to do" rather than the how (this is for documentation). The subtleties can be expressed with a combination of the name and the arguments. The details are in the documentation.

Does the user need to know about `Notifications` abstraction here to do what they want? I think no.

Reply:
Woul adding an `event` suffix help, e.g. `attach_notification_event` or `attach_notification_event_source`

Reply:
The user needs to understand those concepts and the API calls must be different since they are behaving differently, depending on what you attach. If you attach a notification you cannot expect a timeout and spurious wakeups, with deadlines you can.

I would be open to rename the API but they should be different and clearly state notification, deadline or interval.

Reply:
I still can't resolve that "attach notification" is actually attaching a listener. It requires a couple of cognitive jumps to make sense of. I leave it up to you though as to whether you agree or how much you want to focus on this aspect.

Reply:
The original API comes from rust and there it is

```rust
fn attach_notification<T: SynchronousMultiplexing>(value: &T);
```

In the C++ API a notification can be a listener or a file descriptor - there is already the overload for that since the trait approach in C++ would be messy at best.

Reply:
@elfenpiff Let's leave it for now, but I don't think it should be called `attach_notification` in the Rust API either. A`Listener` or `FileDescriptor` is attached, for which notifications are received.